solutions/cses-2431.mdx

```diff
@@ -0,0 +1,172 @@
+---
+id: cses-2431
+source: CSES
+title: Digit Queries
+author: Nathan Gong
+---
+
+Since $k$ can be as big as $10^{18}$, there is no way we can approach this
+problem through brute force. Instead, we'll need to work out an algorithm that
+can answer queries in logarithmic time.
+
+Let's group numbers by the amount of digits they have. An observation we can
+make is that for any $n \geq 1$, there are $9 \cdot 10^{n-1}$ numbers in the
+group with numbers of length $n$. Additionally, for each group, the amount of
+digits from all of the group's numbers combined is $n$ times the amount of
+numbers in the group. For instance:
+
+- Length $1$ numbers: $1$...$9$ -> $9$ numbers -> $1\cdot9=9$ digits total
+- Length $2$ numbers: $10...99$ -> $90$ numbers -> $2\cdot90=180$ digits total
+- Length $3$ numbers: $100...999$ -> $900$ numbers -> $3\cdot900=2700$ digits
+  total
+- And so on...
+
+To find which group the number that contains the $k$th digit is in, we can
+iterate through groups until the sum of total the number of digits from all
+groups we've processed becomes greater than or equal to $k$. In other words, we
+continue increasing the number of groups we look at (starting from $1$-digit
+numbers) until the following is true:
+
+$$
+\sum^{\text{\# of groups}}_{n=1}{n\cdot9\cdot10^{n-1}\geq k}
+$$
+
+Once the condition becomes true, we know that the $k$th digit must fall into the
+last group used in the sum. From there, we can calculate the exact number that
+the $k$th digit is in, which we can then use to find the value of the $k$th
+digit with modular arithmetic.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(q\log{k})$
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+typedef long long ll;
+
+ll pow10(int exp) {
```

Comment:
think c++ has builtin pow
yeah here: https://en.cppreference.com/w/cpp/numeric/math/pow

Reply:
https://stackoverflow.com/questions/15851636/why-is-my-integer-math-with-stdpow-giving-the-wrong-answer

Not super familiar with c++ library functions, but it works with floating point numbers and apparently can also be inaccurate

Reply:
ah ok
~~add comment please~~

Reply:
yes sir